### PR TITLE
[SPARK-25061][SQL] Precedence for ThriftServer hiveconf commandline parameter

### PR DESCRIPTION
**Pull request by ajithme**
### What changes were proposed in this pull request?
In `HiveClientImpl.scala` when we create new SessionState, we prepare configuration and the options passed to `start-thriftserver.sh` via  `--hiveConf` need to take precedence when creating `HiveConf`  i.e  in order 
`hadoopConf < sparkConf < overrideProps(--hiveconf) < extraConfig`

### Why are the changes needed?
As per the documentation here, https://spark.apache.org/docs/latest/sql-distributed-sql-engine.html user can provide `--hiveconf` to override the hive configurations when using `start-thriftserver.sh` but as per the code, https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L182 here, the hive-site properties (part of hadoopConf) will override the configuration done from command line which is not as per expectation

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Added UT along with Base UTs can Pass

### CODE REVIEW

**Comment by ajithme** [comment]
cc @srowen @dongjoon-hyun @HeartSaVioR

**Comment by srowen** [comment]
I think @yhuai wrote the comment above about why it's processed in that order, so may be able to review better

**Comment by ajithme** [comment]
gentle ping @HyukjinKwon @steveloughran @cloud-fan @yhuai @gatorsmile

**Comment by ajithme** [comment]
gentle ping @yhuai @dongjoon-hyun @HyukjinKwon

**Comment by dongjoon-hyun** [comment]
ok to test

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 1]
```diff
@@ -177,9 +177,14 @@ private[hive] class HiveClientImpl(
     // has hive-site.xml. So, HiveConf will use that to override its default values.
     // 2: we set all spark confs to this hiveConf.
     // 3: we set all entries in config to this hiveConf.
+
+    // not to loose command line overwritten properties
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: According to the context, do you mean `lose` instead of `loose`?

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @ajithme .
In general, we need one of the followings.
1. A unit test case.
2. An explicit reproducible procedure.

In the section `Tested this patch manually` of the PR description, could you elaborate your procedure to verify this PR? Please write what you did with the command lines and show the result of *before* and *after* this PR.

**Comment by dongjoon-hyun** [comment]
Could you check the UT failure, @ajithme ?

**Comment by ajithme** [comment]
> Could you check the UT failure, @ajithme ?

Sure, will update shortly

**Comment by ajithme** [comment]
Found the test case failures due to `org.apache.spark.sql.hive.client.HiveClientImpl#extraConfig` sent by `org.apache.spark.sql.hive.HiveUtils#newTemporaryConfiguration` were lost due to `overriddenHiveProps`. Hence fixing the order i.e 
`hadoopConf < sparkConf < overrideProps< extraConfig` 

Please retest. Will update the PR description with manual steps for verification shortly

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 1 above)
Comment (ajithme) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: updated

**Comment by ajithme** [comment]
@dongjoon-hyun I have updated the PR with testcase failure also added a UT to reproduce and verify the issue. Please review

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 2]
```diff
@@ -177,8 +177,12 @@ private[hive] class HiveClientImpl(
     // has hive-site.xml. So, HiveConf will use that to override its default values.
     // 2: we set all spark confs to this hiveConf.
     // 3: we set all entries in config to this hiveConf.
+
+    // not to lose command line overwritten properties
+    // make a copy overridden props so that it can be reinserted finally
+    val overriddenHiveProps = HiveConf.getConfSystemProperties.asScala
     val confMap = (hadoopConf.iterator().asScala.map(kv => kv.getKey -> kv.getValue) ++
-      sparkConf.getAll.toMap ++ extraConfig).toMap
+      sparkConf.getAll.toMap ++ overriddenHiveProps ++ extraConfig).toMap
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Seems we should update https://github.com/apache/spark/pull/27041/files#diff-6fd847124f8eae45ba2de1cf7d6296feR170-R179 and also explain why extraConfig is at the end.

Comment (ajithme) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: @yhuai Sure, I have updated the PR with reasonable pointers for the order. Does it suffice it now.?

Comment (yhuai) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Thank you. As getConfSystemProperties will get all of hive confs that are in the system properties, it is possible that we will pull in a config that is not set by `--hiveconf`. Seems we are introducing a behavior change? Can you explain the impact of this change and why this change is fine?

Comment (ajithme) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Sure, even without my changes in this PR, the `HiveConf` always considers the hive confs in system properties which were not set via `--hiveconf` as part of `HiveConf` constructor i.e Refer https://github.com/apache/hive/blob/rel/release-2.3.5/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java#L4079 (same behaviour for hive 1.2.1 as well) hence, this do not change any flow

**Comment by ajithme** [comment]
gentle ping @dongjoon-hyun @yhuai

**Comment by ajithme** [comment]
gentle ping @dongjoon-hyun @yhuai @cloud-fan @HyukjinKwon can we get this fix in 3.0.?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 3]
```diff
@@ -176,9 +176,18 @@ private[hive] class HiveClientImpl(
...
+    // 3: we take the conf passed as --hiveconf which would be set as system properties
+    // by org.apache.hive.service.server.HiveServer2.ServerOptionsProcessor.parse in
+    // org.apache.spark.sql.hive.thriftserver.HiveThriftServer2.main.
+    // 4: we set all entries in extraConfig to this hiveConf which have the highest precedence.
+    // To summarize, the order of precedence will be
+    // hadoopConf < sparkConf < overrideProps < extraConfig
+
+    // not to lose command line overwritten properties
+    // make a copy overridden props so that it can be reinserted finally
+    val overriddenHiveProps = HiveConf.getConfSystemProperties.asScala
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: so we totally ignore the --hive-conf previously?

Comment (ajithme) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: it was handled as --hiveconf as part of HiveConf constructor i.e Refer https://github.com/apache/hive/blob/rel/release-2.3.5/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java#L4079 , 
i.e first it loads hive-site and then it adds --hiveconf  properties on top. 

But in spark we again add hadoopConf on top of it hence overwriting HiveConf order

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Can you add more code comment to convince people that `HiveConf.getConfSystemProperties` contains only the --hiveconf?

Comment (ajithme) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: @cloud-fan updated. is the comment adequate now?

**Comment by cloud-fan** [comment]
@wangyum @yaooqinn can you take a look?

**Comment by yaooqinn** [comment]
Thanks for pinging me @cloud-fan

We can override hive configurations in many ways.

Take `hive.metastore.uris` for an example, we can reset it via `--conf spark.hadoop.hive.metastore.uris=thrift://example.com:9083`, or `--conf spark.hive.hive.metastore.uris=thrift://example.com:9083` or `--hiveconf hive.metastore.uris=thrift://example.com:9083` or maybe(not sure) `--conf spark.driver.extraJavaOptions=-Dxxx`. 

This PR seems to prefer `--hiveconf` than others.

Personally, I prefer those spark configurations always have higher precedence than other type
configurations including hive/hadoop/java/system, etc, as we are writing spark applicaitons.

**Comment by ajithme** [comment]
> Thanks for pinging me @cloud-fan
> 
> We can override hive configurations in many ways.
> 
> Take `hive.metastore.uris` for an example, we can reset it via `--conf spark.hadoop.hive.metastore.uris=thrift://example.com:9083`, or `--conf spark.hive.hive.metastore.uris=thrift://example.com:9083` or `--hiveconf hive.metastore.uris=thrift://example.com:9083` or maybe(not sure) `--conf spark.driver.extraJavaOptions=-Dxxx`.
> 
> This PR seems to prefer `--hiveconf` than others.
> 
> Personally, I prefer those spark configurations always have higher precedence than other type
> configurations including hive/hadoop/java/system, etc, as we are writing spark applications.

Thanks @yaooqinn for your thoughts. This seems little confusing to know who is overriding as per the documentation mentioned in https://spark.apache.org/docs/latest/sql-distributed-sql-engine.html#running-the-thrift-jdbcodbc-server 

I agree with your opinion of having sparkConf as most precedence, but command line (--hiveconf) should be preferred over config file (hive-site.xml)

For the case you mentioned (X marks the conf is used)

type | case 1 | case 2 | case 3 | case 4
------------ | ------------- | ------------- | ------------- | -------------
--conf spark.hadoop.hive.* | X | X | - | -
--conf spark.hive.hive.* | X | - | X | -
--hiveconf  | X | X | X | X
hive-site.xml  | X | X | X | X
**Preference** | ? | --conf spark.hadoop.hive.* | --conf spark.hive.* | --hiveconf

so do you mean, in case 1, 2, 3 where spark.* conf is used it must get preference.?
how about case 4.?

i prefer in case 4 --hiveconf has precedence and rest cases spark conf can have higher precedence

**Comment by yaooqinn** [comment]
I have no stong option about case 4.

BTW, you  should also pay attention to `SparkSQLCLIDriver` where the hive `sesstionState` is initialized before the spark context

**Pull request closed without merging**